File: pathplanner/finder/messaging.py
from typing import Callable, Iterable, NamedTuple, Optional, Sequence, Tuple


class SearchStep(NamedTuple):
    step_index: int
    current: tuple[int, int]
    visited: tuple[tuple[int, int], ...]
    open_nodes: tuple[tuple[int, int], ...]
    blocked: tuple[tuple[int, int], ...]


SearchStepCallback = Optional[Callable[[SearchStep], None]]


def normalize_cells(cells: Iterable[tuple[int, int]]) -> tuple[tuple[int, int], ...]:
    return tuple(sorted(set(cells)))


def collect_open_nodes(
    openq: Sequence[tuple[float, tuple[int, int]]],
    visited: Iterable[tuple[int, int]],
) -> tuple[tuple[int, int], ...]:
    visited_set = set(visited)
    return tuple(node for _, node in openq if node not in visited_set)


def emit_step(
    on_step: SearchStepCallback,
    step_index: int,
    current: tuple[int, int],
    visited: Iterable[tuple[int, int]],
    openq: Sequence[tuple[float, tuple[int, int]]],
    blocked: Iterable[tuple[int, int]],
) -> None:
    if not on_step:
        return

    visited_tuple = normalize_cells(visited)
    open_nodes = collect_open_nodes(openq, visited_tuple)
    blocked_tuple: Tuple[tuple[int, int], ...] = tuple(sorted(blocked))

    on_step(
        SearchStep(
            step_index=step_index,
            current=current,
            visited=visited_tuple,
            open_nodes=open_nodes,
            blocked=blocked_tuple,
        )
    )

File: rustfinder/src/finder/messaging.rs
use std::collections::BTreeSet;

#[derive(Debug, Clone)]
pub struct SearchStep {
    pub step_index: usize,
    pub current: (usize, usize),
    pub visited: Vec<(usize, usize)>,
    pub open_nodes: Vec<(usize, usize)>,
    pub blocked: Vec<(isize, isize)>,
}

pub type SearchStepCallback<'a> = Option<&'a mut dyn FnMut(SearchStep)>;

fn normalize_cells<I>(cells: I) -> Vec<(usize, usize)>
where
    I: IntoIterator<Item = (usize, usize)>,
{
    let mut set: BTreeSet<(usize, usize)> = BTreeSet::new();
    for cell in cells {
        set.insert(cell);
    }
    set.into_iter().collect()
}

fn collect_open_nodes(
    open_nodes: &[(usize, usize)],
    visited: &[(usize, usize)],
) -> Vec<(usize, usize)> {
    let visited_set: BTreeSet<(usize, usize)> = visited.iter().copied().collect();
    open_nodes
        .iter()
        .copied()
        .filter(|node| !visited_set.contains(node))
        .collect()
}

pub fn emit_step(
    callback: &mut SearchStepCallback,
    step_index: usize,
    current: (usize, usize),
    visited: impl IntoIterator<Item = (usize, usize)>,
    open_nodes: &[(usize, usize)],
    blocked: &[(isize, isize)],
) {
    let Some(handler) = callback else {
        return;
    };

    let visited_vec = normalize_cells(visited);
    let open_filtered = collect_open_nodes(open_nodes, &visited_vec);
    let mut blocked_vec: Vec<(isize, isize)> = blocked.to_vec();
    blocked_vec.sort();

    handler(SearchStep {
        step_index,
        current,
        visited: visited_vec,
        open_nodes: open_filtered,
        blocked: blocked_vec,
    });
}
